swift/apple/FirezoneKit/Sources/FirezoneKit/Features/MainView.swift

```diff
@@ -15,70 +15,47 @@ import SwiftUI
   final class MainViewModel: ObservableObject {
     private let logger: AppLogger
     private var cancellables: Set<AnyCancellable> = []
+    let tunnelStore: TunnelStore
 
-    let appStore: AppStore
     @Dependency(\.mainQueue) private var mainQueue
 
-    @Published var loginStatus: AuthStore.LoginStatus = .uninitialized
-    @Published var tunnelStatus: NEVPNStatus = .invalid
-    @Published var orderedResources: [DisplayableResources.Resource] = []
+    @Published private(set) var resources: [Resource]?
 
-    init(appStore: AppStore) {
-      self.appStore = appStore
-      self.logger = appStore.logger
+    init(tunnelStore: TunnelStore, logger: AppLogger) {
+      self.tunnelStore = tunnelStore
+      self.logger = logger
       setupObservers()
     }
 
     private func setupObservers() {
-      appStore.authStore.$loginStatus
-        .receive(on: mainQueue)
-        .sink { [weak self] loginStatus in
-          self?.loginStatus = loginStatus
-        }
-        .store(in: &cancellables)
-
-      appStore.tunnelStore.$status
+      tunnelStore.$status
         .receive(on: mainQueue)
         .sink { [weak self] status in
-          self?.tunnelStatus = status
+          guard let self = self else { return }
           if status == .connected {
-            self?.appStore.tunnelStore.beginUpdatingResources()
+            self.tunnelStore.beginUpdatingResources()
           } else {
-            self?.appStore.tunnelStore.endUpdatingResources()
+            self.tunnelStore.endUpdatingResources()
           }
         }
         .store(in: &cancellables)
```

Comment:
> It's very different from Rust. Did I understand it correctly?

Spot on! I just learned a couple things from this too.

> ChatGPT told me that the main queue will drop old events or use backpressure to avoid using an unbounded amount of RAM if the subscribers can't process events fast enough. I'm not sure if both or either of those things are true?

Probably true. I would trust ChatGPT's "intuition" on this, because it's one of the most common concepts in Swift UI programming.